src/hotspot/share/logging/circularStringBuffer.hpp

```diff
@@ -0,0 +1,212 @@
+#define SHARE_LOGGING_CIRCULARSTRINGBUFFER_HPP
+#include "logging/logFileStreamOutput.hpp"
+#include "nmt/memTracker.hpp"
+#include "runtime/mutex.hpp"
+#include "runtime/os.inline.hpp"
+#include "runtime/semaphore.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/resourceHash.hpp"
+
+#include <string.h>
+#ifdef LINUX
+#include <sys/mman.h>
+#endif
+
+// The CircularMapping is a struct that provides
+// an interface for writing and reading bytes in a circular buffer
+// correctly. This indirection is necessary because there are two
+// underlying implementations: Linux, and all others.
+#ifdef LINUX
+// Implements a circular buffer by using the virtual memory mapping facilities of the OS.
+// Specifically, it reserves virtual memory with twice the size of the requested buffer.
+// The latter half of this buffer is then mapped back to the start of the first buffer.
+// This allows for write_bytes and read_bytes to consist of a single memcpy, as the
+// wrap-around is dealt with by the virtual memory system.
+struct CircularMapping {
+  FILE* file;
+  char* buffer;
+  size_t size;
+
+  CircularMapping()
+  : file(nullptr), buffer(nullptr), size(0) {
+  };
+
+  CircularMapping(size_t size);
+  ~CircularMapping() {
+    munmap(buffer, size * 2);
+    fclose(file);
+  }
+
+  void write_bytes(size_t at, const char* bytes, size_t size) {
+    memcpy(&buffer[at], bytes, size);
+  }
+
+  void read_bytes(size_t at, char* out, size_t size) {
+    memcpy(out, &buffer[at], size);
+  }
+};
+#else
+struct CircularMapping {
+  char* buffer;
+  size_t size;
+  CircularMapping()
+  : buffer(nullptr), size(0) {
+  };
+  CircularMapping(size_t size);
+
+  void write_bytes(size_t at, const char* bytes, size_t size) {
+    const size_t part1_size = MIN2(size, this->size - at);
+    const size_t part2_size = size - part1_size;
+
+    ::memcpy(&buffer[at], bytes, part1_size);
+    ::memcpy(buffer, &bytes[part1_size], part2_size);
+  }
+
+  void read_bytes(size_t at, char* out, size_t size) {
+    const size_t part1_size = MIN2(size, this->size - at);
+    const size_t part2_size = size - part1_size;
+
+    ::memcpy(out, &buffer[at], part1_size);
+    ::memcpy(&out[part1_size], buffer, part2_size);
+  }
+
+  ~CircularMapping() {
+    os::release_memory(buffer, size);
+  }
+};
+#endif
+
+class CircularStringBuffer {
+  friend class AsyncLogTest;
+
+public:
+    // account for dropped messages
+  using StatisticsMap = ResourceHashtable<LogFileStreamOutput*, uint32_t, 17, /*table_size*/
+                                        AnyObj::C_HEAP, mtLogging>;
+private:
+  static const LogDecorations& None;
+  const bool _should_stall; // Should a producer stall until a consumer has made room for its message?
+
+  // Need to perform accounting of statistics under a separate lock.
+  StatisticsMap& _stats;
+  PlatformMonitor& _stats_lock;
+
+  // Can't use a Monitor here as we need a low-level API that can be used without Thread::current().
```

Comment:
Thank you for finding the original comment thread. Wouldn't it be sufficient to add a `TBIVM` if and only if `Thread::current_or_null() != nullptr`?

Reply:
That will mean a little duplicated code due to the TBIVM being block-scoped (which we do elsewhere). The TBIVM is  necessary but I'm not certain it is sufficient to deal with all safepoint issues.

Reply:
From reading TBIVM, I think I'm using it incorrectly. It's currently this:

```c++
      if (thread != nullptr && thread->is_Java_thread()) {
        ThreadBlockInVM tbivm(JavaThread::cast(thread));
        while (not_enough_memory()) {
          _producer_lock.wait(0);
        }
      }
```

But maybe it should be this:

```c++
      if (thread != nullptr && thread->is_Java_thread()) {
        while (not_enough_memory()) {
        ThreadBlockInVM tbivm(JavaThread::cast(thread));
          _producer_lock.wait(0);
        }
      }
```

`Monitor` is also a bit confusing to me, it seems that only `JavaThread`s can use `Monitor` (see `Monitor::wait` for example), but we have no guarantee that UL is used from a `JavaThread`.

At the very least, this is revealing that I'm not that good at the JVM's thread and locking system :-).

Reply:
Right, I talked to ErikÖ, and the first version (the current one in the source tree) is correct. The second one isn't wrong either, but it's unnecessary.

I'm not creative enough to see any other safepoint issues, however.

Reply:
Yeah the in-or-out-of the loop placement of TBIVM only impacts performance/overhead.

`Monitor::wait` includes safepoint checks and so is only for JavaThreads. For NonJavaTHreads and JavaThread's that don't want to perform safepoint checks you use `Monitor::wait_without_safepoint_check`. This used to be handled by an argument passed to `wait` but the API was reworked a few releases back. The caller has to check what kind of thread they have. So you will need to restructure the wait code you have.

Other safepoint issues would relate to potential deadlocks if the UL code is called whilst other locks are held and a safepoint is active. For non-stalling locking the short critical sections make that a non-issue, but with stalling it could well be an issue.

Reply:
>Other safepoint issues would relate to potential deadlocks if the UL code is called whilst other locks are held and a safepoint is active. For non-stalling locking the short critical sections make that a non-issue, but with stalling it could well be an issue.

You mean a non-issue in practice, but not in theory, right? If we do have a reachable deadlock state, then it doesn't matter if UL stalls or not, it  just increases the probability of that state being reached, is this what you mean? I'm having some trouble constructing the state you're talking about in my head. Stalling only occurs for a log site, this is not relevant for the log writer. Therefore, there is a progress guarantee on the log sites in the sense that at least one log site will always progress. This would change if the consumer thread was unable to progress, due to a safepoint-induced deadlock for example.


As an aside:  There is a starvation risk, as the lock is not fair. There's also the issue of a big log message being unable to progress, as it's stalling for memory but smaller log sites are able to progress, thus preventing it from ever getting enough memory. I don't actually think that these are problems in practice, as we typically do have moments of quiescence, the throughput is in the higher MiB/s and the log buffer is in the multiple MiBs in size.

Reply:
As we discussed off-list the starvation issue is a real problem that needs to be addressed.

Overall the memory management for async UL seems somewhat problematic and adding to the complexity in the stalling case. But that is probably a topic for another day.